File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/presentation/ui/pilih_db/PilihDbFragment.java
package com.bits.bee.bpmc.presentation.ui.pilih_db;

import java.lang.System;

/**
 * Created by aldi on 04/03/22.
 */
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000d\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0010\u000b\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0003\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0005\n\u0002\u0010\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0005\b\u0007\u0018\u00002\b\u0012\u0004\u0012\u00020\u00020\u0001B)\u0012\"\b\u0002\u0010\u0003\u001a\u001c\u0012\u0004\u0012\u00020\u0005\u0012\u0006\u0012\u0004\u0018\u00010\u0006\u0012\u0004\u0012\u00020\u0007\u0012\u0004\u0012\u00020\u00020\u0004\u00a2\u0006\u0002\u0010\bJ\b\u0010\u0015\u001a\u00020\u0016H\u0016J\u0018\u0010\u0017\u001a\u00020\u00162\u0006\u0010\u0018\u001a\u00020\u00192\u0006\u0010\u001a\u001a\u00020\u001bH\u0016J\u0010\u0010\u001c\u001a\u00020\u00072\u0006\u0010\u001d\u001a\u00020\u001eH\u0016J\u001a\u0010\u001f\u001a\u00020\u00162\u0006\u0010 \u001a\u00020!2\b\u0010\"\u001a\u0004\u0018\u00010#H\u0016J\u0010\u0010$\u001a\u00020\u00162\u0006\u0010%\u001a\u00020\u0007H\u0002J\b\u0010&\u001a\u00020\u0016H\u0016J\b\u0010\'\u001a\u00020\u0016H\u0016R\u000e\u0010\t\u001a\u00020\nX\u0082.\u00a2\u0006\u0002\n\u0000R.\u0010\u0003\u001a\u001c\u0012\u0004\u0012\u00020\u0005\u0012\u0006\u0012\u0004\u0018\u00010\u0006\u0012\u0004\u0012\u00020\u0007\u0012\u0004\u0012\u00020\u00020\u0004X\u0096\u0004\u00a2\u0006\b\n\u0000\u001a\u0004\b\u000b\u0010\fR\u000e\u0010\r\u001a\u00020\u000eX\u0082.\u00a2\u0006\u0002\n\u0000R\u001b\u0010\u000f\u001a\u00020\u00108BX\u0082\u0084\u0002\u00a2\u0006\f\n\u0004\b\u0013\u0010\u0014\u001a\u0004\b\u0011\u0010\u0012\u00a8\u0006("}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/pilih_db/PilihDbFragment;", "Lcom/bits/bee/bpmc/presentation/base/BaseFragment;", "Lcom/bits/bee/bpmc/databinding/FragmentPilihDbBinding;", "bindingInflater", "Lkotlin/Function3;", "Landroid/view/LayoutInflater;", "Landroid/view/ViewGroup;", "", "(Lkotlin/jvm/functions/Function3;)V", "adapter", "Lcom/bits/bee/bpmc/presentation/ui/pilih_db/PilihDbAdapter;", "getBindingInflater", "()Lkotlin/jvm/functions/Function3;", "dialog", "Lcom/bits/bee/bpmc/presentation/dialog/LoadingDialogHelper;", "viewModel", "Lcom/bits/bee/bpmc/presentation/ui/pilih_db/PilihDbViewModel;", "getViewModel", "()Lcom/bits/bee/bpmc/presentation/ui/pilih_db/PilihDbViewModel;", "viewModel$delegate", "Lkotlin/Lazy;", "initComponents", "", "onCreateOptionsMenu", "menu", "Landroid/view/Menu;", "inflater", "Landroid/view/MenuInflater;", "onOptionsItemSelected", "item", "Landroid/view/MenuItem;", "onViewCreated", "view", "Landroid/view/View;", "savedInstanceState", "Landroid/os/Bundle;", "setVisibilityComponent", "isLoading", "subscribeListeners", "subscribeObservers", "app_debug"})
@dagger.hilt.android.AndroidEntryPoint()
public final class PilihDbFragment extends com.bits.bee.bpmc.presentation.base.BaseFragment<com.bits.bee.bpmc.databinding.FragmentPilihDbBinding> {
    @org.jetbrains.annotations.NotNull()
    private final kotlin.jvm.functions.Function3<android.view.LayoutInflater, android.view.ViewGroup, java.lang.Boolean, com.bits.bee.bpmc.databinding.FragmentPilihDbBinding> bindingInflater = null;
    private final kotlin.Lazy viewModel$delegate = null;
    private com.bits.bee.bpmc.presentation.ui.pilih_db.PilihDbAdapter adapter;
    private com.bits.bee.bpmc.presentation.dialog.LoadingDialogHelper dialog;
    
    public PilihDbFragment() {
        super();
    }
    
    public PilihDbFragment(@org.jetbrains.annotations.NotNull()
    kotlin.jvm.functions.Function3<? super android.view.LayoutInflater, ? super android.view.ViewGroup, ? super java.lang.Boolean, com.bits.bee.bpmc.databinding.FragmentPilihDbBinding> bindingInflater) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    @java.lang.Override()
    public kotlin.jvm.functions.Function3<android.view.LayoutInflater, android.view.ViewGroup, java.lang.Boolean, com.bits.bee.bpmc.databinding.FragmentPilihDbBinding> getBindingInflater() {
        return null;
    }
    
    private final com.bits.bee.bpmc.presentation.ui.pilih_db.PilihDbViewModel getViewModel() {
        return null;
    }
    
    @java.lang.Override()
    public void onViewCreated(@org.jetbrains.annotations.NotNull()
    android.view.View view, @org.jetbrains.annotations.Nullable()
    android.os.Bundle savedInstanceState) {
    }
    
    @java.lang.Override()
    public void initComponents() {
    }
    
    @java.lang.Override()
    public void subscribeListeners() {
    }
    
    @java.lang.Override()
    public void subscribeObservers() {
    }
    
    @java.lang.Override()
    public void onCreateOptionsMenu(@org.jetbrains.annotations.NotNull()
    android.view.Menu menu, @org.jetbrains.annotations.NotNull()
    android.view.MenuInflater inflater) {
    }
    
    @java.lang.Override()
    public boolean onOptionsItemSelected(@org.jetbrains.annotations.NotNull()
    android.view.MenuItem item) {
        return false;
    }
    
    /**
     * Check visibiliy of progress bar and recycler view when request data from api
     * @param isLoading
     */
    private final void setVisibilityComponent(boolean isLoading) {
    }
}

File: app/src/main/java/com/bits/bee/bpmc/presentation/ui/pilih_db/PilihDbFragment.kt
package com.bits.bee.bpmc.presentation.ui.pilih_db

import android.Manifest
import android.content.ActivityNotFoundException
import android.content.Intent
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.provider.Settings
import android.view.*
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.fragment.app.viewModels
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.bits.bee.bpmc.R
import com.bits.bee.bpmc.data.data_source.remote.response.LoginResponse
import com.bits.bee.bpmc.databinding.FragmentPilihDbBinding
import com.bits.bee.bpmc.presentation.base.BaseFragment
import com.bits.bee.bpmc.presentation.dialog.DialogBuilderHelper
import com.bits.bee.bpmc.presentation.dialog.LoadingDialogHelper
import com.bits.bee.bpmc.presentation.dialog.NoInternetDialogBuilder
import com.bits.bee.bpmc.presentation.dialog.download.DownloadDialogBuilder
import com.bits.bee.bpmc.utils.BeePreferenceManager
import com.bits.bee.bpmc.utils.Resource
import com.bits.bee.bpmc.utils.extension.gone
import com.bits.bee.bpmc.utils.extension.visible
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch

/**
 * Created by aldi on 04/03/22.
 */

@AndroidEntryPoint
class PilihDbFragment constructor(
    override val bindingInflater: (LayoutInflater, ViewGroup?, Boolean) -> FragmentPilihDbBinding = FragmentPilihDbBinding::inflate
) : BaseFragment<FragmentPilihDbBinding>() {

    //View Model
    private val viewModel : PilihDbViewModel by viewModels()

    private lateinit var adapter: PilihDbAdapter

    private lateinit var dialog : LoadingDialogHelper

    private val requestPermissionCamera = registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()){ data ->
        data.forEach {
            if(!it.value){
//                Toast.makeText(requireActivity(), "Beberapa permission belum aktif!", Toast.LENGTH_LONG).show()
//                try {
//                    val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
//                    val uri = Uri.fromParts("package", "com.bits.bee.bpmc", null)
//                    intent.data = uri
//                    startActivity(intent)
//                }catch (e : ActivityNotFoundException){
//                    e.printStackTrace()
//                }
            }
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        arguments?.let {
            viewModel.state.inputEmail = it.getString("username", "")
            viewModel.state.inputPassword = it.getString("password", "")

            viewModel.login()
        }
    }

    override fun initComponents() {
        dialog = LoadingDialogHelper(requireContext())
        binding.apply {
            setHasOptionsMenu(true)
            adapter = PilihDbAdapter(mListener = object :  PilihDbAdapter.PilihDbAdapterI{
                override fun onItemClick(db: LoginResponse.Db) {
                    viewModel.postDb(db.dbName)
                }
            })
            rvList.layoutManager = LinearLayoutManager(requireContext())
            rvList.adapter = adapter
            findNavController().currentBackStackEntry?.savedStateHandle?.getLiveData<Boolean>("isFinsihDownload")?.observe(viewLifecycleOwner){
                if(it)
                    viewModel.onSuccessDb()
            }
        }
        val permission = mutableListOf(Manifest.permission.READ_EXTERNAL_STORAGE, Manifest.permission.WRITE_EXTERNAL_STORAGE,)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
            permission.add(Manifest.permission.MANAGE_EXTERNAL_STORAGE, )
            permission.add(Manifest.permission.ACCESS_MEDIA_LOCATION)
        }
        requestPermissionCamera.launch(permission.toTypedArray())
    }

    override fun subscribeListeners() {

    }

    override fun subscribeObservers() {
        viewModel.observeLoginResponse().removeObservers(viewLifecycleOwner)
        viewModel.observeLoginResponse().observe(viewLifecycleOwner) {
            it?.let {
                when (it.status) {
                    Resource.Status.LOADING -> {
                        setVisibilityComponent(true)
                    }
                    Resource.Status.SUCCESS -> {
                        setVisibilityComponent(false)
                        it.data?.let {
                            BeePreferenceManager.saveToPreferences(requireActivity(), getString(R.string.pref_email), viewModel.state.inputEmail)
                            adapter.submitList(it.db)
                            if(it.db.isNotEmpty() && it.db.size == 1){
                                viewModel.postDb(it.db[0].dbName)
                            }
                        }
                    }
                    Resource.Status.ERROR -> {
                        setVisibilityComponent(false)
                    }
                    Resource.Status.NOINTERNET -> {
                        val dialog = NoInternetDialogBuilder({
                            viewModel.login()
                        })
                        dialog.show(parentFragmentManager, "")
                    }
                }
            }
        }

        viewModel.observeLicense().removeObservers(viewLifecycleOwner)
        viewModel.observeLicense().observe(viewLifecycleOwner){
            when(it.status){
                Resource.Status.LOADING -> {
                    viewModel.showDialog("Validasi Lisensi")
                }
                Resource.Status.SUCCESS -> {
                    viewModel.hideDialog()
                    it.data?.let {
                        if(it.licNumber.isNotEmpty()){
                            viewModel.onSuccessLicense()
                        } else {
                            val dialog = DialogBuilderHelper.showDialogChoice(requireActivity(), "Informasi",
                                "Lisensi telah digunakan semua!\nUlangi proses verifikasi lisensi?",
                                "Ulangi",
                                {
                                    it.dismiss()
                                    viewModel.postDb(viewModel.state.dbName)
                                },
                                "Tidak"
                            )
                            dialog.show(parentFragmentManager, "")
                        }
                    }
                }
                Resource.Status.ERROR -> {
                    viewModel.hideDialog()
                    val dialog = DialogBuilderHelper.showDialogChoice(requireActivity(), "Informasi",
                        "Lisensi telah digunakan semua!\nUlangi proses verifikasi lisensi?",
                        "Ulangi",
                        {
                            it.dismiss()
                            viewModel.postDb(viewModel.state.dbName)
                        },
                        "Tidak"
                    )
                    dialog.show(parentFragmentManager, "")
                }
                Resource.Status.NOINTERNET -> {
                    val dialog = NoInternetDialogBuilder({
                        viewModel.postDb(viewModel.state.dbName)
                    })
                    dialog.show(parentFragmentManager, "")
                }
            }
        }

        viewLifecycleOwner.lifecycleScope.launch {
            viewLifecycleOwner.repeatOnLifecycle(Lifecycle.State.STARTED){
                viewModel.event.collect {
                    when(it){
                        PilihDbViewModel.UIEvent.NavigateToPilihMode -> {
                            val action = PilihDbFragmentDirections.actionPilihDbFragmentToModeTampilanFragment()
                            findNavController().navigate(action)
                        }
                        PilihDbViewModel.UIEvent.RequestDb -> {

                        }
                        is PilihDbViewModel.UIEvent.ShowLoading -> {
                            val isShow = it.isShow
                            if(isShow)
                                dialog.show(message = it.msg)
                            else
                                dialog.hide()
                        }
                        PilihDbViewModel.UIEvent.NavigateToDownload -> {
                            val dialog = DownloadDialogBuilder(isCancel = false, onFinishDownload = {
                                viewModel.onSuccessDb()
                            })
                            dialog.show(parentFragmentManager, "")
                        }
                        PilihDbViewModel.UIEvent.ShowNoInternet -> {
                            val dialog = NoInternetDialogBuilder({
                                viewModel.postDb(viewModel.state.dbName)
                            })
                            dialog.show(parentFragmentManager, "")
                        }
                    }
                }
            }
        }
        viewLifecycleOwner.lifecycleScope.launch {
            viewModel.viewStates().collect {
                it?.let {

                }
            }
        }

    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.menu_reload, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when(item.itemId){
            R.id.menu_refresh -> viewModel.login()
        }
        return super.onOptionsItemSelected(item)
    }

    /**
     * Check visibiliy of progress bar and recycler view when request data from api
     * @param isLoading
     */
    private fun setVisibilityComponent(isLoading : Boolean){
        binding.apply {
            if(isLoading){
                rvList.gone()
                progressBar.visible()
            } else {
                rvList.visible()
                progressBar.gone()
            }
        }
    }

}